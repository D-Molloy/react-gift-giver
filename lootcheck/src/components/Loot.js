import React, {Component} from 'react';
import { connect } from 'react-redux'
import { fetchBitcoin } from '../actions/bitcoin';

export class Loot extends Component {

    componentDidMount(){
        this.props.fetchBitcoin()
    }


    computeBitcoin(){
        //destructuring bitcoin of the props
        const { bitcoin } = this.props;

        //want to return an empty value if our bitcoin object is empty
        //if we have to return something in our default case, this should be an empty object
        //this is called a guard clause -- if it doesn't meet criteria than we return out of the function
        //used then the state is initalized with no value
        if (Object.keys(bitcoin).length === 0) return '';
        //replacing all of the commas with empty strings so we can parseInt
        return this.props.balance / parseInt(bitcoin.bpi.USD.rate.replace(',',''), 10);


    }

    showPrice(){
        const { bitcoin } = this.props;
        if (Object.keys(bitcoin).length === 0) return '';
        return bitcoin.bpi.USD.rate;
    }

    render(){
        return (
            <div>
                <h3 id="BCPrice">Current Bitcoin price: ${this.showPrice()}</h3>
                <h3 id="BCBalance"># of Bitcoins you can afford: {this.computeBitcoin()}</h3>
            </div>
        )
    }
}

// CONNECT() TAKES TWO PARAMETERS
//first - a function that describes what part of the redux store we want to use on this component
//second - what actionCreators we want to use in the component to send data to the redux store (not currently using right now)
//first - mapping the whole state (balance and bitcoin) to this component
//second - link the action you want to use in the component
export default connect(state => state, {fetchBitcoin})(Loot);