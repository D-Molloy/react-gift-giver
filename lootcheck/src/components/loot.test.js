import React from 'react';
import { shallow, mount } from 'enzyme';
import { Loot } from './Loot';

describe("Loot", ()=>{
    
    const props = { balance: 10, bitcoin: {}}
    let loot = shallow(<Loot {...props}/>);
    

    it("renders properly", ()=>{
        expect(loot).toMatchSnapshot();
    })

    //testing componentDidMount on the fetchBitcoin action
    //need to use `mount` from enzyme
    describe("when mounted", ()=>{
        let mockFetchBitcoin = jest.fn();

        beforeEach(()=>{
            props.fetchBitcoin = mockFetchBitcoin;
            loot = mount(<Loot {...props} />)
        })

        it('dispatches the `fetchBitcoin()` method it receives from props', ()=>{
            expect(mockFetchBitcoin).toHaveBeenCalled();
        })
    })

    describe("when there are valid bitcoin Balance", ()=>{
        beforeEach(()=>{
            let props= {balance: 10, bitcoin: {bpi: { USD: {rate:'1,000'}}}};
            loot = shallow(<Loot {...props} />)
        })
        
        it('displays the correct bitcoin value', ()=> {
            expect(loot.find('#BCBalance').text()).toEqual('# of Bitcoins you can afford: 0.01');
        })

        // it('displays the correct bitcoin Price', ()=> {
        //     expect(loot.find('#BCPrice').text()).toEqual('Bitcoin balance: 0.01');
        // })
        
    })
})