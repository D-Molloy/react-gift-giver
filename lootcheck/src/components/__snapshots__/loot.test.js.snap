// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loot renders properly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Loot
    balance={10}
    bitcoin={Object {}}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h3
          id="BCBalance"
>
          # of Bitcoins you can afford: 
          
</h3>,
        <h3
          id="BCPrice"
>
          Current Bitcoin price: $
          
</h3>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "# of Bitcoins you can afford: ",
            "",
          ],
          "id": "BCBalance",
        },
        "ref": null,
        "rendered": Array [
          "# of Bitcoins you can afford: ",
          "",
        ],
        "type": "h3",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Current Bitcoin price: $",
            "",
          ],
          "id": "BCPrice",
        },
        "ref": null,
        "rendered": Array [
          "Current Bitcoin price: $",
          "",
        ],
        "type": "h3",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h3
            id="BCBalance"
>
            # of Bitcoins you can afford: 
            
</h3>,
          <h3
            id="BCPrice"
>
            Current Bitcoin price: $
            
</h3>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "# of Bitcoins you can afford: ",
              "",
            ],
            "id": "BCBalance",
          },
          "ref": null,
          "rendered": Array [
            "# of Bitcoins you can afford: ",
            "",
          ],
          "type": "h3",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Current Bitcoin price: $",
              "",
            ],
            "id": "BCPrice",
          },
          "ref": null,
          "rendered": Array [
            "Current Bitcoin price: $",
            "",
          ],
          "type": "h3",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
